README.md
[![Build Status](https://travis-ci.org/HsoiEnterprises/HEAlert.svg)](https://travis-ci.org/HsoiEnterprises/HEAlert)
![Cocoapods Compatible](https://img.shields.io/cocoapods/v/HEAlert.svg)
![License](https://img.shields.io/badge/license-BSD%203--Clause-blue.svg)
![Platform](https://img.shields.io/badge/platform-iOS-lightgrey.svg)


# HEAlert
A convenience around `UIAlertController`, because sometimes you just want to say "OK".


# Background and Approach

`UIAlertController` is awesome, but sometimes it's too much. How many times do you want to simply alert with a message and an "OK" button? Somewhat easy with `UIAlertView`, certainly cumbersome with `UIAlertController`, but simple with `HEAlert`.

That's why I created `HEAlert`.

Go from:

```swift
  let alertView = UIAlertView(title: "Hello World", message: nil, delegate: nil, cancelButtonTitle: "OK")
  alertView.show()
```

or:

```swift
let alertController = UIAlertController(title: "Hello World", message: nil, preferredStyle: .Alert)
let action = UIAlertAction(title: "OK", style: .Default, handler: nil)
alertController.addAction(action)
self.presentViewController(alertController, animated: true, completion: nil)
```

**to:**

```swift
HEAlert.ok("Hello World")
```

or:

```swift
HEAlert.ok("Hello World", presentingViewController: self)
```

# Supported OS and SDK

- Xcode 8.0+
- Swift 3+
- iOS 8 (minimum, required)

With v1.0, `HEAlert` has moved to Swift 3 and Xcode 8. If you need need Swift 2.2 support, use the `v0.5.3` tag.  There has been no explicit effort to support Swift 2.3.

With v0.5, `HEAlert` has moved to Swift 2 and Xcode 7. If you still need Xcode 6.x (Swift 1.2) support, you can use older versions of `HEAlert` such as v0.4.2.


# Installation

Currently supports being installed by simple source code addition, by git submodule, or by Cocoapods (preferred).


## Submodule

You can easily obtain git as a submodule, and simply add `HEAlert.swift` to your project. Off you go!

## Cocoapods

[CocoaPods][CocoaPods] is a dependency manager for Cocoa projects.

CocoaPods 0.36 adds supports for Swift and embedded frameworks. You can install it with the following command:

```bash
$ gem install cocoapods
```

To integrate HEAlert into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '8.0'
use_frameworks!

pod 'HEAlert'
```

Then, run the following command:

```bash
$ pod install
```


# Usage

Usage is fairly straightforward, and the code is fully documented. There's also an example application that demonstrates a simple usage of the class.

## Simple OK alert

```swift
HEAlert.ok("Hello World")
```

or:

```swift
HEAlert.ok("Hello World", presentingViewController: self)
```

## OK with Message

```swift
HEAlert.ok("Hello World", message: "This is my first HEAlert", presentingViewController: self)
```

## More Complicated

You can customize a fair portion of the `HEAlert` behaviors:

```swift
HEAlert.other("Do you like waffles?", message: "This is an important question", buttonTitle: "Yes", cancelTitle: "No", presentingViewController: self, cancelHandler: { (_) -> Void in
    println("Must like pancakes instead")
}) { (_) -> Void in
    println("Get the syrup!")
}
```

## NSError

Easy display of an `NSError`:

```swift
HEAlert.error(anNSError, presentingViewController: self, buttonHandler: nil)
```

# Contact

## Hsoi Enterprises
- [Website][hsoienterprises-website]
- [Facebook][hsoienterprises-facebook]
- [Twitter (@HsoiEnterprises)][hsoienterprises-twitter]

## Creator
- John C. Daub [(@hsoi)][hsoi-twitter]


# Changes / Release Notes

See included `CHANGELOG.md` file.


# License

BSD 3-clause “New” or “Revised” License. See included "License" file.


[hsoienterprises-website]: http://www.hsoienterprises.com
[hsoienterprises-facebook]: https://www.facebook.com/HsoiEnterprises
[hsoienterprises-twitter]: http://twitter.com/hsoienterprises
[hsoi-twitter]: http://twitter.com/hsoi
[cocoapods]: http://cocoapods.org


CHANGELOG.md
# HEAlert CHANGELOG

## v1.0.0 - 2016-09-23

Going 1.0, because it's time.

* Brings Swift 3 and Xcode 8 support (thank you @kreeger)

## v0.5.3 - 2016-09-23

**THIS IS THE LAST VERSION TO SUPPORT SWIFT 2.2**

If you need Swift 2.2 support, use this tag. Going forward, `HEAlert` will be Swift 3.

* Fixed Travis for Xcode 7.3


## v0.5.2 - 2016-03-26

* Verified against Xcode 7.3 and Swift 2.2 (no changes made, just verified)
* HEAlert inherits from `NSObject` so it can be bridged into Objective-C code
* Fixed codesigning (thanx to @jlandon)

## v0.5.1 - 2015-09-12

* Minor adjustments for Xcode 7 and Cocoapods.


## v0.5 - 2015-09-12

**NB: Now requires Xcode 7 and Swift 2.**

### Added

* Project file, to enable stand-alone building of HEAlert.framework
* travis-ci support

### Changed

* Conformance with Xcode 7 and Swift 2
* Merged stand-alone sample project into the main .xcodeproj file as a secondary target.


## v0.4.2 - 2015-06-24

### Added

* public declarations

### Fixed

* cleaned up copyright notice

### Improved

* README
* inline docs


## v0.4.1 - 2015-06-23

### Fixed

* Some stray README notes


## v0.4 - 2015-06-23

### Added

* Cocoapods support


## v0.3 - 2015-06-25

### Changed

* Expanding HEAlert to add cancel mechanisms and further flesh out the API.


## v0.2 - 2015-05-12

### Added

* `ok(title, message)`


## v0.1 - 2015-05-08

Initial release.

`HEAlert` existed prior to this, but things brought a conversion to Swift and public release.


HEAlert/HEAlert.h
//
//  HEAlert.h
//
//  Created by jaysonlane on 7/23/15.
//
//  HEAlert - Copyright (c) 2015, Hsoi Enterprises LLC
//  All rights reserved.
//  hsoi@hsoienterprises.com
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of HEAnalytics nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import <UIKit/UIKit.h>
 
FOUNDATION_EXPORT double HEAlertVersionNumber;
FOUNDATION_EXPORT const unsigned char HEAlertVersionString[];

HEAlertTests/HEAlertTests.swift
//
//  HEAlertTests.swift
//  HEAlertTests
//
//  Created by hsoi on 9/5/15.
//  Copyright © 2015 Hsoi Enterprises LLC. All rights reserved.
//

import XCTest
@testable import HEAlert

class HEAlertTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


HEAlertExample/AppDelegate.swift
//
//  AppDelegate.swift
//  HEAlertExample
//
//  Created by hsoi on 5/8/15.
//  Copyright (c) 2015 Hsoi Enterprises LLC. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) ->  Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



HEAlertExample/ViewController.swift
//
//  ViewController.swift
//  HEAlertExample
//
//  Created by hsoi on 5/8/15.
//  Copyright (c) 2015 Hsoi Enterprises LLC. All rights reserved.
//

import UIKit
import HEAlert

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


    @IBAction func showAlert(_ sender: AnyObject) {
        HEAlert.ok("This Is An Alert", presentingViewController: self) { (_) -> Void in
            HEAlert.other("Another Alert", message: "This alert is comes from the buttonHandler of the first alert", buttonTitle: "Cool!", presentingViewController: self)
        }
    }    
}



Example/HEAlertExampleTests/HEAlertExampleTests.swift
//
//  HEAlertExampleTests.swift
//  HEAlertExampleTests
//
//  Created by hsoi on 5/8/15.
//  Copyright (c) 2015 Hsoi Enterprises LLC. All rights reserved.
//

import UIKit
import XCTest

class HEAlertExampleTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


HEAlert/HEAlert.swift
//
//  HEAlert.swift
//
//  Created by hsoi on 2010-12-24, converted to Swift 2015-05-08
//
//  HEAnalytics - Copyright (c) 2010, 2015-2016, Hsoi Enterprises LLC
//  All rights reserved.
//  hsoi@hsoienterprises.com
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  * Neither the name of HEAlert nor the names of its
//  contributors may be used to endorse or promote products derived from
//  this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import Foundation
import UIKit


/**
HEAlert - a utility class to simplify the presentation of alerts.

UIAlertController is excellent, but the complexity can be a bit much when all you need is a simple "OK" alert.
HEAlert wraps up UIAlertController, providing a simplified but expansive API to make alerts easier.

Go from this:

```
let alertController = UIAlertController(title: "Hello World", message: nil, preferredStyle: .Alert)
let action = UIAlertAction(title: "OK", style: .Default, handler: nil)
alertController.addAction(action)
self.presentViewController(alertController, animated: true, completion: nil)
```

to this:

```
HEAlert.ok("Hello World", presentingViewController: self)
```
*/
open class HEAlert: NSObject {

    /**
    Presents an "OK" alert with the given title.

    NB: While this is the easist function to use, you should strive to avoid using it and prefer to use functions that
    take a presentingViewController. This is not an ideal and forward-looking mechanism, but is provided because sometimes
    it is what you need to solve an immediate issue.

    - parameter title:                       The title of the alert.
    */
    open class func ok(_ title: String) {
        HEAlert.showAlert(title, message: nil, buttonTitle: nil, cancelTitle: nil, presentingViewController: nil, cancelHandler: nil, buttonHandler: nil)
    }



    /**
    Presents an "OK" alert with the given title and message.

    NB: While this is the easist function to use, you should strive to avoid using it and prefer to use functions that
    take a presentingViewController. This is not an ideal and forward-looking mechanism, but is provided because sometimes
    it is what you need to solve an immediate issue.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message.
    */
    open class func ok(_ title: String, message: String) {
        HEAlert.showAlert(title, message: message, buttonTitle: nil, cancelTitle: nil, presentingViewController: nil, cancelHandler: nil, buttonHandler: nil)
    }



    /**
    Presents an "OK" alert with the given title, upon the given presentingViewController.

    Provides for an optional closure to be executed when the OK button is tapped.

    - parameter title:                       The title of the alert.
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func ok(_ title: String, presentingViewController: UIViewController, buttonHandler: ((UIAlertAction?) -> Void)? = nil) {
        HEAlert.showAlert(title, message: nil, buttonTitle: nil, cancelTitle: nil, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Presents an "OK" alert with the given title and message, upon the given presentingViewController.

    Provides for an optional closure to be executed when the OK button is tapped.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message.
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func ok(_ title: String, message: String, presentingViewController: UIViewController, buttonHandler: ((UIAlertAction?) -> Void)? = nil) {
        HEAlert.showAlert(title, message: message, buttonTitle: nil, cancelTitle: nil, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Presents an alert with the given title and custom button title, upon the given presentingViewController.

    Provides for an optional closure to be executed when the button is tapped.

    - parameter title:                       The title of the alert.
    - parameter buttonTitle:                 The button's title.
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func other(_ title: String, buttonTitle: String, presentingViewController: UIViewController, buttonHandler: ((UIAlertAction?) -> Void)? = nil) {
        HEAlert.showAlert(title, message: nil, buttonTitle: buttonTitle, cancelTitle: nil, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Presents an alert with the given title, message, and custom button title, upon the given presentingViewController.

    Provides for an optional closure to be executed when the button is tapped.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message.
    - parameter buttonTitle:                 The button's title.
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func other(_ title: String, message: String, buttonTitle: String, presentingViewController: UIViewController, buttonHandler: ((UIAlertAction?) -> Void)? = nil) {
        HEAlert.showAlert(title, message: message, buttonTitle: buttonTitle, cancelTitle: nil, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Presents an alert with the given title, message, custom Cancel button title, and custom Default button title; upon the given presentingViewController.

    Often you need 2 buttons, one to do something and one to not do something (but still dismiss the alert). So the cancelTitle allows for a Cancel button
    (or whatever you wish to name it "Later", "No", etc.), and the buttonTitle allows for the "OK"/default button and then the optional buttonHandler,
    if provided, will be executed upon the tap to that OK/default button.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message.
    - parameter buttonTitle:                 The button's title.
    - parameter cancelTitle:                 The cancel button's title
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the default button.
    */
    open class func other(_ title: String, message: String, buttonTitle: String, cancelTitle: String, presentingViewController: UIViewController, buttonHandler: ((UIAlertAction?) -> Void)?) {
        HEAlert.showAlert(title, message: message, buttonTitle: buttonTitle, cancelTitle: cancelTitle, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Presents an alert with the given title, message, custom Cancel button title, and custom Default button title; upon the given presentingViewController.

    Often you need 2 buttons, one to do something and one do something else. So the cancelTitle allows for a Cancel button (or whatever you wish
    to name it "Later", "No", etc.) and a cancelHandler to be invoked upon tapping that button, and the buttonTitle allows for the "OK"/default button
    and then the optional buttonHandler will be executed, if provided, upon the tap to that OK/default button.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message.
    - parameter buttonTitle:                 The button's title.
    - parameter cancelTitle:                 The cancel button's title
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter cancelHandler:               An optional closure invoked in response to the user tapping the cancel button.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func other(_ title: String, message: String, buttonTitle: String, cancelTitle: String, presentingViewController: UIViewController, cancelHandler: ((UIAlertAction?) -> Void)?, buttonHandler: ((UIAlertAction?) -> Void)?) {
        HEAlert.showAlert(title, message: message, buttonTitle: buttonTitle, cancelTitle: cancelTitle, presentingViewController: presentingViewController, cancelHandler: cancelHandler, buttonHandler: buttonHandler)
    }



    /**
    Presents an "OK" alert for the given NSError, upon the given presentingViewController.

    The alert's title will be the localizedDescription, and the alert's message will be the localizedRecoverySuggestion.

    Provides for an optional closure to be executed when the "OK" button is tapped.

    - parameter error:                       The NSError to display.
    - parameter presentingViewController:    The UIViewController that will present the alert.
    - parameter buttonHandler:               An optional closure invoked in response to the user tapping the button.
    */
    open class func error(_ error: NSError, presentingViewController: UIViewController?, buttonHandler: ((UIAlertAction?) -> Void)? = nil) {
        HEAlert.showAlert(error.localizedDescription, message: error.localizedRecoverySuggestion, buttonTitle: nil, cancelTitle: nil, presentingViewController: presentingViewController, cancelHandler: nil, buttonHandler: buttonHandler)
    }



    /**
    Private function that actually does all of the heavy lifting.

    Note that for cancelTitle and buttonHandler to be honored, there must be a presentingViewController. We need the presentingViewController so we
    can have the hook to execute the buttonHandler upon tapping the main/default/OK button. And if we cannot excute some action in response to
    tapping main/default/OK button, there's no point in having an alternative "cancel" button that will do nothing.

    - parameter title:                       The title of the alert.
    - parameter message:                     The alert message. If no message is provided, no message is displayed.
    - parameter buttonTitle:                 The alert has a single button (default/OK), and this will be the title. If nil or empty, will default to "OK".
    - parameter cancelTitle:                 The cancel button's title. If nil or empty, will not be present (a 1 button alert). Will only be utilized if presentingViewController != nil
    - parameter presentingViewController:    The UIViewController that will present the UIAlertController. Preferred to be non-nil, but if nil will cause logic to fall back to UIAlertView (this may change depending upon Apple's future directions)
    - parameter cancelHandler:               Optional closure invoked in response to the user tapping the cancel button. Only utilized if presentingViewController != nil and cancelTitle is non-nil and non-empty. BUT perfectly acceptable to have a valid cancelTitle and a nil cancelHandler (maybe you don't want to do anything interesting upon cancel).
    - parameter buttonHandler:               Optional closure invoked in response to the user tapping the button. Only utilized if presentingViewController != nil.
    */
    private class func showAlert(_ title: String, message: String?, buttonTitle: String?, cancelTitle: String?, presentingViewController: UIViewController?, cancelHandler: ((UIAlertAction?) -> Void)?, buttonHandler: ((UIAlertAction?) -> Void)?) {

        // Hsoi 2015-05-08 - One useful thing about UIAlertView is you can present them from anywhere -- you don't
        // need a presenting UIViewController. While some may argue you should never show an alert from
        // a place where you don't have access to a UIViewController (e.g. from within model code), sometimes
        // reality or necessity trumps ideals and you need this ability.
        //
        // This is why the presentingViewController is optional. If you want you CAN pass nil, meaning
        // we don't have a UIViewController, but we still want to present an alert. I chose to NOT make
        // it have a default value of nil tho, so it's clear in code this is an explicit choice
        // and not just something we perhaps forgot to type.
        //
        // For now, if presentingViewController is nil, we will fall back to UIAlertView. I filed
        // rdar://problem/18297508 back in 2014-09-10 requesting some way to still be able to present
        // alerts without need for a presentingViewController, but as of this writing there's been no
        // comment or movement on it (and honestly I'm not expecting Apple will do anything about this).
        // One alternative implementation is to get the window's rootViewController, but I don't think
        // that's ideal (what if you have more than 1 window?)
        //
        // Ideally, you should pass a presentingViewController -- and this is why most of the HEAlert API makes
        // the presentingViewController non-optional. However, we contend for it, and just realize that if you
        // do use a nil presentingViewController that it may have issues, may be less than ideal, and may have to
        // go away someday and you'll then have to figure out how to (re)solve the issue into having a
        // UIViewController to present from -- thus probably best to just solve that problem now.

        var okTitle: String = NSLocalizedString("OK", comment: "Alert button title")
        if let buttonTitle = buttonTitle , !buttonTitle.isEmpty {
            okTitle = buttonTitle
        }


        if let presentingViewController = presentingViewController {
            let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)

            let okAction = UIAlertAction(title: okTitle, style: .default, handler: buttonHandler)
            alertController.addAction(okAction)

            if let cancelTitle = cancelTitle , !cancelTitle.isEmpty {
                let cancelAction = UIAlertAction(title: cancelTitle, style: .cancel, handler: cancelHandler)
                alertController.addAction(cancelAction)
            }

            presentingViewController.present(alertController, animated: true, completion: nil)
        }
        else {
            assert(buttonHandler == nil, "buttonHandler will not be executed if there's no presentingViewController")
            assert(cancelTitle == nil, "cancelTitle will not be honored since there's no way to handle a buttonHandler and thus have 2 distinct results from 2 different button taps")

            let alertView = UIAlertView(title: title, message: message, delegate: nil, cancelButtonTitle: okTitle)
            alertView.show()
        }
    }
}
